File: web/src/types/maintenance.ts
export type HealthBand = "healthy" | "fair" | "attention" | "unknown";
export type Severity = "info" | "warning" | "critical";
export type Confidence = "low" | "medium" | "high";
export type ScanFrequency = "off" | "hourly" | "daily";
export type Preset = "conservative" | "balanced" | "aggressive";

export interface HealthFactor {
  score: number;
  value?: number | string | null;
  detail: string;
  weight: number;
}

export interface HealthSummary {
  score: number | null;
  band: HealthBand;
  table_count: number;
  attention_count: number;
  total_data_bytes: number;
}

export interface TableHealth {
  schema_name: string;
  table_name: string;
  score: number | null;
  band: HealthBand;
  scanned_at: string | null;
  snapshot_count: number | null;
  data_file_count: number | null;
  manifest_count: number | null;
  total_data_bytes: number | null;
  avg_file_bytes: number | null;
  small_file_ratio: number | null;
  orphan_bytes: number | null;
  factors: Record<string, HealthFactor> | null;
}

export interface WorkspaceHealth {
  workspace_id: string;
  slug: string;
  summary: HealthSummary;
}

export interface DeploymentHealth {
  summary: HealthSummary;
  workspaces: WorkspaceHealth[];
}

export interface NamespaceHealth {
  schema_name: string;
  summary: HealthSummary;
}

export interface WorkspaceHealthDetail {
  summary: HealthSummary;
  namespaces: NamespaceHealth[];
  tables: TableHealth[];
}

export interface HealthHistoryPoint {
  scanned_at: string;
  score: number | null;
  total_data_bytes: number | null;
}

export interface Remediation {
  applicable_in_app: boolean;
  summary?: string;
  command?: string;
  tool?: string;
  warning?: string;
}

export interface Recommendation {
  id: string;
  workspace_id: string;
  schema_name: string;
  table_name: string;
  kind: string;
  severity: Severity;
  confidence: Confidence;
  rationale: string;
  estimated_impact: Record<string, number | string | null> | null;
  remediation: Remediation | null;
  status: string;
  created_at: string;
  resolved_at: string | null;
}

export interface TableHealthDetail {
  table: TableHealth;
  history: HealthHistoryPoint[];
  recommendations: Recommendation[];
}

export interface MaintenancePolicy {
  scan_enabled: boolean;
  scan_frequency: ScanFrequency;
  preset: Preset;
  thresholds: Record<string, number>;
  max_tables_per_cycle: number;
  last_scan_at: string | null;
  last_deep_scan_at: string | null;
}

export interface PolicyUpdate {
  scan_enabled?: boolean;
  scan_frequency?: ScanFrequency;
  preset?: Preset;
  thresholds?: Record<string, number>;
  max_tables_per_cycle?: number;
}

export interface ScanResult {
  status: string;
  dispatched: number;
  candidates: number;
  stale: number;
  deep: boolean;
}

File: api/src/api/schemas/maintenance.py
"""Response/request shapes for the maintenance advisor API."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class PolicyOut(BaseModel):
    scan_enabled: bool
    scan_frequency: str
    preset: str
    thresholds: dict
    max_tables_per_cycle: int
    last_scan_at: datetime | None = None
    last_deep_scan_at: datetime | None = None


class PolicyUpdate(BaseModel):
    scan_enabled: bool | None = None
    scan_frequency: str | None = None
    preset: str | None = None
    # Advanced override of individual threshold values; merged onto the preset.
    thresholds: dict | None = None
    max_tables_per_cycle: int | None = None


class ScanResult(BaseModel):
    status: str
    dispatched: int = 0
    candidates: int = 0
    stale: int = 0
    deep: bool = False


class HealthSummary(BaseModel):
    """A rolled-up score for any scope (namespace, workspace, deployment)."""

    score: int | None
    band: str
    table_count: int
    attention_count: int
    total_data_bytes: int


class TableHealthOut(BaseModel):
    schema_name: str
    table_name: str
    score: int | None
    band: str
    scanned_at: datetime | None = None
    snapshot_count: int | None = None
    data_file_count: int | None = None
    manifest_count: int | None = None
    total_data_bytes: int | None = None
    avg_file_bytes: int | None = None
    small_file_ratio: float | None = None
    orphan_bytes: int | None = None
    factors: dict | None = None


class NamespaceHealthOut(BaseModel):
    schema_name: str
    summary: HealthSummary


class WorkspaceHealthOut(BaseModel):
    workspace_id: uuid.UUID
    slug: str
    summary: HealthSummary


class DeploymentHealthOut(BaseModel):
    summary: HealthSummary
    workspaces: list[WorkspaceHealthOut]


class WorkspaceHealthDetailOut(BaseModel):
    summary: HealthSummary
    namespaces: list[NamespaceHealthOut]
    tables: list[TableHealthOut]


class HealthHistoryPoint(BaseModel):
    scanned_at: datetime
    score: int | None = None
    total_data_bytes: int | None = None


class RecommendationOut(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    schema_name: str
    table_name: str
    kind: str
    severity: str
    confidence: str
    rationale: str
    estimated_impact: dict | None = None
    remediation: dict | None = None
    status: str
    created_at: datetime
    resolved_at: datetime | None = None


class TableHealthDetailOut(BaseModel):
    table: TableHealthOut
    history: list[HealthHistoryPoint]
    recommendations: list[RecommendationOut]
